### ISSUE TITLE
User-defined mappings of event to seen values

### ISSUE THREAD

**Issue by julienfantin** [Day8/re-frame-async-flow-fx#6]
We need async event coordination during the application lifecycle to track different remote fxs for the same page.

`async-flow-fx` does almost exactly what we need, except the current implementation can't  distinguish `[:post-success 123]` from `[:post-success 456]` which will lead to race conditions for our use-case.

The culprit here seems to be that the [async event handler](https://github.com/Day8/re-frame-async-flow-fx/blob/master/src/day8/re_frame/async_flow_fx.cljs#L120) hard-codes tracking `seen` events by their `event-id`.

My first idea was to extend the fx to track specific events, like:

```clojure
{:first-dispatch
 [:do-X 123]
 :rules [{:when :seen? :events [:success-X 123] :dispatch [:do-Y 123]}
         {:when     :seen-any-of?
          :events   [[:fail-X 123] [:fail-Y 123] [:fail-Z 123]]
          :dispatch [:app-failed-state] :halt? true}]}
```

However some async events may not be known at the time the fx is declared, what if `:fail-Y` is dispatched as `[:fail-Y {:reason :timeout} 123]`, we can't specify that in advance.

We need a more dynamic solution that will let us define the logic that will map an event to its `seen` value, on a per-event-id basis. I think the following would do what we want:

```clojure
{:event->seen
 { ;; here we would dispatch [:failX 123] which is what we want to track
  :failX identity
  ;; here we would dispatch [:failY {:reason :timeout} 123] but track [:failY 123] as its seen value
  :failY (fn [[event-id _ entity-id]] [event-id entity-id])}
 :first-dispatch :do-X
 :rules          [{:when :seen? :events :success-X :dispatch :do-Y}
                  {:when     :seen-any-of?
                   :events   [[:fail-X 123] [:fail-Y 123] :fail-Z]
                   :dispatch [:app-failed-state] :halt? true}]}
```

Here `event->seen` is an optional map of event id to a fn of `event -> seen value`. It's also simple to maintain backwards compatibility by defaulting to the `first` fn if there is no `event->seen` fn for an event id.

**Comment by julienfantin** [comment]
Updated the issue description, a couple of questions:

Do you think this is a valid use-case for the library? And if so, do you think the proposed solution addresses the issue?

**Comment by danielcompton** [comment]
This is interesting, and not an approach we really considered. Our re-frame apps generally distinguish semantically between different values, e.g. `[:post-success 123]`, `[:post-error 456]`. Are your two events semantically the same apart from their later values? I guess what I'm asking is, does it make more sense for your app to handle those two values differently by calling different handlers?

**Comment by julienfantin** [comment]
No, the events both apply to the same type, but have a separate identity, determined by their args.

Let's say we have two remote calls required to load a `foo` : `[:fetch-meta-data <foo/id>]`, `[:fetch-content <foo/id>]` with matching success/error callbacks. The events all apply to the same `foo` type but each event address a specific unique `foo` by `<foo/id>`.

So the goal here is to load multiple `foo`s concurrently by running multiple instances of the same `foo` flow for different values of `<foo/id>`. This is possible because the response events also contain `<foo/id>` allowing us to match them with the events that originated them

**Comment by julienfantin** [comment]
Actually hacked it together in a fork that I haven't pushed yet, here is what the modified forwarding test looks like:

```clojure
(deftest test-forwarding-event->seen
  (let [flow {:event-id->event->seen {:1 identity
                                      :3 identity
                                      :4 identity}
              :first-dispatch [:start]
              :id             :test-id
              :db-path        [:p]
              :rules [{:id 0 :when :seen? :events [[:1 1]] :dispatch [:2]}
                      {:id 1 :when :seen? :events [[:3 3]] :halt? true}
                      {:id 2 :when :seen-any-of? :events [[:4 4] :5] :dispatch [:6]}]}
        handler-fn  (core/make-flow-event-handler flow)]

    ;; event :no should cause nothing to happen
    (is (= (handler-fn
            {:db {:p {:seen-events #{:33}
                      :rules-fired #{}}}}
            [:test-id [:no]])
           {:db {:p {:seen-events #{:33 :no}
                     :rules-fired #{}}}}))

    ;; new event should not cause a new dispatch because task is already started  (:id 0 is in :rules-fired)
    (is (= (handler-fn
            {:db {:p {:seen-events #{[:1 1]}
                      :rules-fired #{0}}}}
            [:test-id [:1 1]])
           {:db {:p {:seen-events #{[:1 1]} :rules-fired #{0}}}}))

    ;; new event should cause a dispatch
    (is (= (handler-fn
            {:db {:p {:seen-events #{}
                      :rules-fired #{}}}}
            [:test-id [:1 1]])
           {:db {:p {:seen-events #{[:1 1]} :rules-fired #{0}}}
            :dispatch-n (list [:2])}))

    ;; new event should cause a dispatch
    (is (= (handler-fn
            {:db {:p {:seen-events #{[:1 1]}
                      :rules-fired #{0}}}}
            [:test-id [:3 3]])
           {:db {:p {:seen-events #{[:1 1] [:3 3]} :rules-fired #{0 1}}}
            :dispatch-n (list [:test-id :halt-flow])}))

    ;; make sure :seen-any-of? works
    (is (= (handler-fn
            {:db {:p {:seen-events #{}
                      :rules-fired #{}}}}
            [:test-id [:4 4]])
           {:db {:p {:seen-events #{[:4 4]} :rules-fired #{2}}}
            :dispatch-n (list [:6])}))))
```

**Comment by danielcompton** [comment]
> So the goal here is to load multiple foos concurrently by running multiple instances of the same foo flow for different values of <foo/id>.

Great, I understand completely now. This is definitely something we'd want to support, feel free to open a PR with your changes and we can take a look.

**Comment by julienfantin** [comment]
Ok great, will integrate in our app first and I'll submit the PR once we've ironed out any issues.

**Comment by julienfantin** [comment]
Forgot to mention we need this in our JVM so we have to convert the project to cljc, as well as `re-frame-forward-events-fx`. Though the conversions are trivial, I wanted to check with you if it's ok to make that part of the PR?

**Comment by danielcompton** [comment]
Yeah sure thing.

**Comment by stumitchell** [comment]
This looks useful would be great to see that pr

**Comment by julienfantin** [comment]
So I implemented what I initially proposed: https://github.com/vimsical/re-frame-async-flow-fx/commit/efcdce7e9ba9211b290a55f572524788677c5588

However with usage I realized this is not a great API.

Our use case is to coordinate a pair of events: 
- `[:data-store-foo-success {:entity/id 123} entity-123-foo-data]` 
- `[:data-store-bar-success {:entity/id 123} entity-123-bar-data]` 

And dispatch a finale event:
- `[:all-data-stores-success {:entity/id 123} entity-123-foo-data entity-123-bar-data]`.

With the implemented solution we're able to correctly define the flow using only data that's know in advance:
- a rule looks like `{:when :seen-both #{[:data-store-foo-success {:entity/id 123}] [:data-store-foo-success {:entity/id 123}]}` 
- an `event-id->seen-fn` map where the seen-fn (of `event -> seen value`) ignores the results `(fn [[id entity _ignore-result]] [id entity])` for both `:data-store-foo-success` and `:data-store-bar-success`

Even though this works, it requires that we store the results in the db for both `:data-store-foo-success` and `:data-store-bar-success`, just to pull them out again in the final event that the rule dispatches.

I realized what we _really_ want is to be able to provide a fn for the dispatch, which would take a map of `{:<event-id> original-event}`. This way we wouldn't need to store the intermediary results, and we could handle all the async reponses in the final handler. However the catch is `original-event`, meaning we'd need to keep track of both the original event, as well as its `seen` value.

We've implemented the sub-par solution in our app and it works well event though it's not elegant due to the temporary storage locations. I was thinking of looking into implementing the proper solution when we get around to refactoring, which may be a little while now.

In the meantime I'm more than happy to hear suggetions on the API/implementation or have someone take it from here.

**Comment by stumitchell** [comment]
@julienfantin can you have a look at https://github.com/Day8/re-frame-async-flow-fx/pull/8 and see if it fits your needs.

**Comment by julienfantin** [comment]
@stumitchell this mostly replicates the functionality in our fork with a simpler implementation so I like it. 

The use-case we had in the end is still unresolved however. To reiterate,  we want rules to have a  dynamic dispatch based the union of their event vectors.

For example if we're waiting for responses `[:success-a 123 resp-a]` and `[:success-b 123 resp-b]` to finally dispatch `[:success-a-and-b 123 resp-a resp-b]` right now we need to use the app-db as temporary storage. 

We'd really like to be able to specify ruls with a fn for `:dispatch` that'd be invoked in the above example with `(dispatch-fn {:success-a [:success-a 123 resp-a] :success-b [:success-b 123 resp-b]})`

But this PR is a step in the right direction since `seen-events` now properly captures the full event vectors, so I could see that feature being submitted as a separate PR that builds on top of this.

**Comment by hipitihop** [comment]
@julienfantin We have been very tardy here while busy on other things, sorry, but we have now also merged #20 does that constitute the last mile to your use case and if so are you happy to close this ?

**Comment by julienfantin** [comment]
@hipitihop ah yes indeed, the combination of the events predicate and dispatch fn now covers this use-case 👍
